staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,759 @@
+
+        self.restake();
+    }
+
+    /// Unstakes the given amount from the inner account of the predecessor.
+    /// The inner account should have enough staked balance.
+    /// The new total unstaked balance will be available for withdrawal in 3 epochs.
+    pub fn unstake(&mut self, amount: U128) {
+        self.ping();
+
+        let amount: Balance = amount.into();
+        assert!(amount > 0, "Unstaking amount should be positive");
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+
+        assert!(
+            self.total_staked_balance > 0,
+            "The contract doesn't have staked balance"
+        );
+        // Calculate the number of shares required to unstake the given amount.
+        // NOTE: The number of shares the account will pay is rounded up, to avoid giving extra
+        // amount.
+        let num_shares = self.num_shares_from_amount_rounded_up(amount);
+        assert!(
+            account.stake_shares >= num_shares,
+            "Not enough staked balance to unstake"
+        );
+        account.stake_shares -= num_shares;
+        account.unstaked += amount;
+        account.unstaked_available_epoch_height = env::epoch_height() + NUM_EPOCHS_TO_UNLOCK;
+        self.save_account(&account_id, &account);
+
+        self.total_staked_balance -= amount;
+        self.total_stake_shares -= num_shares;
+
+        env::log(
+            format!(
+                "@{} unstaking {}. Spent {} staking shares. Total {} unstaked balance and {} staking shares",
+                account_id, amount, num_shares, account.unstaked, account.stake_shares
+            )
+                .as_bytes(),
+        );
+        env::log(
+            format!(
+                "Contract total staked balance is {}. Total number of shares {}",
+                self.total_staked_balance, self.total_stake_shares
+            )
+            .as_bytes(),
+        );
+
+        self.restake();
+    }
+
+    /// Restakes the current `total_staked_balance` again.
+    ///
+    /// NOTE: The staking action may arrive on the next epoch, which means this account might
+    /// accumulate a reward. To avoid skipping this reward we first call `ping` function to
+    /// distribute reward and the inner state.
+    /// Once the staking action happened we need to update the inner balances by calling
+    /// `internal_after_stake`.
+    fn restake(&mut self) {
+        Promise::new(env::current_account_id())
+            .function_call(b"ping".to_vec(), b"{}".to_vec(), 0, PING_GAS)
+            .stake(self.total_staked_balance, self.stake_public_key.clone())
+            .function_call(
+                b"internal_after_stake".to_vec(),
+                b"{}".to_vec(),
+                0,
+                INTERNAL_AFTER_STAKE_GAS,
+            );
+    }
+
+    /// Private method to be called after stake action to update inner balances.
+    pub fn internal_after_stake(&mut self) {
```

Comment:
it is not private?

Reply:
It has to be called by the contract as an action, so it's private in sense of `current == predecessor`